client-v2/src/main/java/com/clickhouse/client/api/Client.java

```diff
@@ -159,29 +181,137 @@ public boolean ping(int timeout) {
+        if (settings == null) {
+            settings = new InsertSettings();
+        }
+
+        boolean hasDefaults = this.hasDefaults.get(data.get(0).getClass());
+        if (hasDefaults) {
+            settings.setFormat(ClickHouseFormat.RowBinaryWithDefaults);
+        } else {
+            settings.setFormat(ClickHouseFormat.RowBinary);
+        }
+
+
+        //Create an output stream to write the data to
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+
         //Lookup the Serializer for the POJO
+        List<POJOSerializer> serializers = this.serializers.get(data.get(0).getClass());
+        if (serializers == null || serializers.isEmpty()) {
+            throw new IllegalArgumentException("No serializer found for the given class. Please register() before calling this method.");
+        }
+
+        long s2 = System.currentTimeMillis();
         //Call the static .serialize method on the POJOSerializer for each object in the list
-        return null;//This is just a placeholder
+        for (Object obj : data) {
+            for (POJOSerializer serializer : serializers) {
+                serializer.serialize(obj, stream);
+            }
+        }
+        long s3 = System.currentTimeMillis();
+
+        LOG.debug("Total serialization time: {}", s3 - s1);
+        return insert(tableName, new ByteArrayInputStream(stream.toByteArray()), settings);
     }
 
     /**
      * Insert data into ClickHouse using a binary stream
      */
-    public Future<InsertResponse> insert(String tableName,
+    public InsertResponse insert(String tableName,
                                      InputStream data,
-                                     InsertSettings settings) throws ClickHouseException, SocketException {
-        return null;//This is just a placeholder
+                                     InsertSettings settings) throws IOException, ExecutionException, InterruptedException {
+        long s1 = System.currentTimeMillis();
+        InsertResponse response;
+        try (ClickHouseClient client = createClient()) {
+            ClickHouseRequest.Mutation request = createMutationRequest(client.write(getServerNode()), tableName, settings)
+                    .format((ClickHouseFormat) settings.getSetting(ClickHouseClientOption.FORMAT.getKey()));
+
+            Future<ClickHouseResponse> future;
+            try(ClickHousePipedOutputStream stream = ClickHouseDataStreamFactory.getInstance().createPipedOutputStream(request.getConfig())) {
+                future = request.data(stream.getInputStream()).execute();
+
+                //Copy the data from the input stream to the output stream
+                byte[] buffer = new byte[8196];
+                int bytesRead;
+                while ((bytesRead = data.read(buffer)) != -1) {
+                    stream.write(buffer, 0, bytesRead);
+                }
+            }
+            response = new InsertResponse(client, future.get());
```

Comment:
this we may do under try catch to handle execution exception that report that operation timeouted to be user friendly, I think